bundles/AdminBundle/Controller/Admin/RecyclebinController.php

```diff
@@ -124,23 +126,91 @@ public function listAction(Request $request)
                 }
             }
 
+            if (!$this->getAdminUser()->isAdmin()) {
+                $conditionFilters[] = $this->getPermittedPaths();
+            }
+
             if (!empty($conditionFilters)) {
                 $condition = implode(' AND ', $conditionFilters);
                 $list->setCondition($condition);
             }
 
+
             $items = $list->load();
             $data = [];
             if (is_array($items)) {
+
                 /** @var Recyclebin\Item $item */
                 foreach ($items as $item) {
-                    $data[] = $item->getObjectVars();
+                    $dataRow = $item->getObjectVars();
+                    $closestParent = $item->getClosestExistingParent();
+                    if ($closestParent instanceof AbstractElement) {
+                        $dataRow['permissions'] = $closestParent->getUserPermissions();
+                    }
+                    $data[] = $dataRow;
                 }
             }
 
             return $this->adminJson(['data' => $data, 'success' => true, 'total' => $list->getTotalCount()]);
         }
     }
+    protected function getPermittedPaths($types = ['asset', 'document', 'object'])
```

Comment:
can we put this in a service so it can be used on `SimpleBackendSearchBundle` and avoid duplication?

Reply:
Can we also may split this in smaller functions?